{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0fd0e8f1",
   "metadata": {},
   "outputs": [
    {
     "ename": "ApplicationError",
     "evalue": "Non-zero return code 127 from 'clustalo -i lncRNA.fa -o aligned.fasta --auto -v', message '/bin/sh: 1: clustalo: not found'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mApplicationError\u001b[0m                          Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_69688/784011622.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     23\u001b[0m \u001b[0;31m# Run Clustal Omega to align the sequences\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     24\u001b[0m \u001b[0mclustalomega_cline\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mClustalOmegaCommandline\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minfile\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfasta_file\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moutfile\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'aligned.fasta'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mverbose\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mauto\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 25\u001b[0;31m \u001b[0mstdout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstderr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mclustalomega_cline\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     26\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     27\u001b[0m \u001b[0;31m# Parse the resulting alignment\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.9/site-packages/Bio/Application/__init__.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, stdin, stdout, stderr, cwd, env)\u001b[0m\n\u001b[1;32m    572\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    573\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mreturn_code\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 574\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mApplicationError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreturn_code\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstdout_str\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstderr_str\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    575\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mstdout_str\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstderr_str\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    576\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mApplicationError\u001b[0m: Non-zero return code 127 from 'clustalo -i lncRNA.fa -o aligned.fasta --auto -v', message '/bin/sh: 1: clustalo: not found'"
     ]
    }
   ],
   "source": [
    "from Bio.Align.Applications import ClustalOmegaCommandline\n",
    "from Bio import AlignIO\n",
    "\n",
    "def calculate_conservation_score(alignment):\n",
    "    \"\"\"Calculate conservation score for each position in a multiple sequence alignment\"\"\"\n",
    "    num_sequences = len(alignment)\n",
    "    seq_length = alignment.get_alignment_length()\n",
    "    conservation_scores = []\n",
    "    for i in range(seq_length):\n",
    "        column = alignment[:, i]\n",
    "        column_residues = [s.upper() for s in column if s != '-']\n",
    "        num_residues = len(column_residues)\n",
    "        if num_residues == 0:\n",
    "            conservation_scores.append(0)\n",
    "        else:\n",
    "            conservation = sum([column_residues.count(s) / num_residues for s in set(column_residues)])\n",
    "            conservation_scores.append(conservation)\n",
    "    return conservation_scores\n",
    "\n",
    "# Path to input fasta file\n",
    "fasta_file = 'lncRNA.fa'\n",
    "\n",
    "# Run Clustal Omega to align the sequences\n",
    "clustalomega_cline = ClustalOmegaCommandline(infile=fasta_file, outfile='aligned.fasta', verbose=True, auto=True)\n",
    "stdout, stderr = clustalomega_cline()\n",
    "\n",
    "# Parse the resulting alignment\n",
    "alignment = AlignIO.read('aligned.fasta', 'fasta')\n",
    "\n",
    "# Calculate conservation score for each sequence in the alignment\n",
    "for record in alignment:\n",
    "    conservation_scores = calculate_conservation_score(alignment[:, record.id])\n",
    "    print(f'{record.id}: {conservation_scores}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "914c2575",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "nthread = 0\n",
      "nthreadpair = 0\n",
      "nthreadtb = 0\n",
      "ppenalty_ex = 0\n",
      "stacksize: 8192 kb\n",
      "generating a scoring matrix for nucleotide (dist=200) ... done\n",
      "Gap Penalty = -1.53, +0.00, +0.00\n",
      "\n",
      "\n",
      "\n",
      "Making a distance matrix ..\n",
      " 3801 / 3828\n",
      "done.\n",
      "\n",
      "Constructing a UPGMA tree (efffree=0) ... \n",
      " 3820 / 3828\n",
      "done.\n",
      "\n",
      "Progressive alignment 1/2... \n",
      "STEP  2201 / 3827 \n",
      "Reallocating..done. *alloclen = 29581\n",
      "STEP  2301 / 3827 \n",
      "Reallocating..done. *alloclen = 34088\n",
      "\n",
      "Reallocating..done. *alloclen = 38550\n",
      "\n",
      "Reallocating..done. *alloclen = 40525\n",
      "\n",
      "len1=30628, len2=8897, Switching to the memsave mode\n",
      "STEP  2401 / 3827 mDP 00001 / 00002DP 00002 / 00002DP 00001 / 0000100001 / 00001DP 00001 / 00001DP 00001 / 00001DP 00001 / 00001DP 00001 / 00001DP 00001 / 00001DP 00001 / 00001DP 00001 / 00001DP 00001 / 00001DP 00001 / 00001DP 00001 / 00001DP 00001 / 00006DP 00002 / 00006DP 00003 / 00006DP 00004 / 00006DP 00005 / 00006DP 00006 / 00006DP 00001 / 00001DP 00001 / 00001DP 00001 / 00001DP 00001 / 00003DP 00002 / 00003DP 00003 / 00003DP 00001 / 00001DP 00001 / 00001DP 00001 / 00001DP 00001 / 00002DP 00002 / 00002DP 00001 / 00001DP 00001 / 00001DP 00001 / 00001DP 00001 / 00006DP 00002 / 00006DP 00003 / 00006DP 00004 / 00006DP 00005 / 00006DP 00006 / 00006DP 00001 / 00001DP 00001 / 00001DP 00001 / 00001DP 00001 / 00001DP 00001 / 00001DP 00001 / 00001DP 00001 / 00001DP 00001 / 00001DP 00001 / 00002DP 00002 / 00002DP 00001 / 00001DP 00001 / 00001DP 00001 / 00001DP 00001 / 00001DP 00001 / 00001DP 00001 / 00004DP 00002 / 00004DP 00003 / 00004DP 00004 / 00004DP 00001 / 00001DP 00001 / 00001DP 00001 / 00001DP 00001 / 00001DP 00001 / 00001DP 00001 / 00004DP 00002 / 00004DP 00003 / 00004DP 00004 / 00004DP 00001 / 00001DP 00001 / 00001DP 00001 / 00001DP 00001 / 00001DP 00001 / 00001"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_69688/3092128952.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     41\u001b[0m     \u001b[0;31m# Run MAFFT to align the sequences\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     42\u001b[0m     \u001b[0mmafft_cline\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msubprocess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mPopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'mafft'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'--auto'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfasta_file\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstdout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msubprocess\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mPIPE\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 43\u001b[0;31m     \u001b[0maligned_sequences\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmafft_cline\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcommunicate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     44\u001b[0m     \u001b[0maligned_fasta_file\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'aligned.fasta'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     45\u001b[0m     \u001b[0;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maligned_fasta_file\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'wb'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.9/subprocess.py\u001b[0m in \u001b[0;36mcommunicate\u001b[0;34m(self, input, timeout)\u001b[0m\n\u001b[1;32m   1119\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_stdin_write\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minput\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1120\u001b[0m             \u001b[0;32melif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstdout\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1121\u001b[0;31m                 \u001b[0mstdout\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstdout\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1122\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstdout\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1123\u001b[0m             \u001b[0;32melif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstderr\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "import os\n",
    "import subprocess\n",
    "\n",
    "def parse_fasta_file(fasta_file):\n",
    "    \"\"\"Parse a FASTA file and return a dictionary of sequences\"\"\"\n",
    "    sequences = {}\n",
    "    with open(fasta_file, 'r') as f:\n",
    "        sequence = ''\n",
    "        seq_name = ''\n",
    "        for line in f:\n",
    "            if line.startswith('>'):\n",
    "                if sequence:\n",
    "                    sequences[seq_name] = sequence\n",
    "                    sequence = ''\n",
    "                seq_name = line[1:].strip()\n",
    "            else:\n",
    "                sequence += line.strip()\n",
    "        sequences[seq_name] = sequence\n",
    "    return sequences\n",
    "\n",
    "def calculate_conservation_score(sequences):\n",
    "    \"\"\"Calculate the conservation score for each position in the alignment\"\"\"\n",
    "    alignment_length = len(list(sequences.values())[0])\n",
    "    num_sequences = len(sequences)\n",
    "    conservation_scores = [0] * alignment_length\n",
    "    for i in range(alignment_length):\n",
    "        column = [seq[i] for seq in sequences.values()]\n",
    "        counts = {}\n",
    "        for char in column:\n",
    "            if char not in counts:\n",
    "                counts[char] = 0\n",
    "            counts[char] += 1\n",
    "        for char, count in counts.items():\n",
    "            frequency = count / num_sequences\n",
    "            conservation_scores[i] += frequency * frequency\n",
    "    return conservation_scores\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    fasta_file = 'lncRNA.fa'\n",
    "    \n",
    "    # Run MAFFT to align the sequences\n",
    "    mafft_cline = subprocess.Popen(['mafft', '--auto', fasta_file], stdout=subprocess.PIPE)\n",
    "    aligned_sequences, err = mafft_cline.communicate()\n",
    "    aligned_fasta_file = 'aligned.fasta'\n",
    "    with open(aligned_fasta_file, 'wb') as f:\n",
    "        f.write(aligned_sequences)\n",
    "    \n",
    "    # Parse the resulting alignment\n",
    "    sequences = parse_fasta_file(aligned_fasta_file)\n",
    "    \n",
    "    # Calculate conservation scores\n",
    "    conservation_scores = calculate_conservation_score(sequences)\n",
    "    \n",
    "    # Print the conservation scores for each position\n",
    "    for i, score in enumerate(conservation_scores):\n",
    "        print(f'Position {i+1}: {score:.4f}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c3bd5032",
   "metadata": {},
   "outputs": [
    {
     "ename": "ApplicationError",
     "evalue": "Non-zero return code 127 from 'clustalo -i lncRNA.fa -o aligned.fasta --auto -v', message '/bin/sh: 1: clustalo: not found'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mApplicationError\u001b[0m                          Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_69688/1780652131.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     52\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     53\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0m__name__\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'__main__'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 54\u001b[0;31m     \u001b[0mmain\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'lncRNA.fa'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     55\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     56\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/tmp/ipykernel_69688/1780652131.py\u001b[0m in \u001b[0;36mmain\u001b[0;34m(fasta_file)\u001b[0m\n\u001b[1;32m     37\u001b[0m     \u001b[0;31m# Align the sequences using Clustal Omega\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     38\u001b[0m     \u001b[0mclustalomega_cline\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mClustalOmegaCommandline\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minfile\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfasta_file\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moutfile\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'aligned.fasta'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mverbose\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mauto\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 39\u001b[0;31m     \u001b[0mstdout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstderr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mclustalomega_cline\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     40\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     41\u001b[0m     \u001b[0;31m# Parse the resulting alignment\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.9/site-packages/Bio/Application/__init__.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, stdin, stdout, stderr, cwd, env)\u001b[0m\n\u001b[1;32m    572\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    573\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mreturn_code\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 574\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mApplicationError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreturn_code\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstdout_str\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstderr_str\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    575\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mstdout_str\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstderr_str\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    576\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mApplicationError\u001b[0m: Non-zero return code 127 from 'clustalo -i lncRNA.fa -o aligned.fasta --auto -v', message '/bin/sh: 1: clustalo: not found'"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import subprocess\n",
    "from Bio import SeqIO\n",
    "from Bio.Align.Applications import ClustalOmegaCommandline\n",
    "\n",
    "def parse_fasta_file(fasta_file):\n",
    "    \"\"\"Parse a FASTA file and return a dictionary of sequences\"\"\"\n",
    "    sequences = {}\n",
    "    with open(fasta_file, 'r') as f:\n",
    "        for record in SeqIO.parse(f, 'fasta'):\n",
    "            sequences[record.id] = str(record.seq)\n",
    "    return sequences\n",
    "\n",
    "def calculate_conservation_score(aligned_sequence):\n",
    "    \"\"\"Calculate the conservation score for an aligned sequence\"\"\"\n",
    "    num_sequences = len(aligned_sequence)\n",
    "    seq_length = len(aligned_sequence[0])\n",
    "    conservation_score = []\n",
    "    for i in range(seq_length):\n",
    "        column = [aligned_sequence[j][i] for j in range(num_sequences)]\n",
    "        num_gaps = column.count('-')\n",
    "        non_gaps = [x for x in column if x != '-']\n",
    "        if len(set(non_gaps)) == 1:\n",
    "            conservation_score.append(1.0)\n",
    "        else:\n",
    "            counts = {}\n",
    "            for nt in non_gaps:\n",
    "                counts[nt] = counts.get(nt, 0) + 1\n",
    "            max_count = max(counts.values())\n",
    "            conservation_score.append(max_count / (num_sequences - num_gaps))\n",
    "    return conservation_score\n",
    "\n",
    "def main(fasta_file):\n",
    "    # Parse the FASTA file\n",
    "    sequences = parse_fasta_file(fasta_file)\n",
    "\n",
    "    # Align the sequences using Clustal Omega\n",
    "    clustalomega_cline = ClustalOmegaCommandline(infile=fasta_file, outfile='aligned.fasta', verbose=True, auto=True)\n",
    "    stdout, stderr = clustalomega_cline()\n",
    "\n",
    "    # Parse the resulting alignment\n",
    "    aligned_sequences = {}\n",
    "    with open('aligned.fasta', 'r') as f:\n",
    "        for record in SeqIO.parse(f, 'fasta'):\n",
    "            aligned_sequences[record.id] = str(record.seq)\n",
    "\n",
    "    # Calculate the conservation score for each sequence\n",
    "    for seq_name, sequence in sequences.items():\n",
    "        aligned_sequence = aligned_sequences[seq_name]\n",
    "        conservation_score = calculate_conservation_score([aligned_sequence])\n",
    "        print(f'{seq_name}: {conservation_score}')\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    main('lncRNA.fa')\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "09e8aad3",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
